src/downsampling_worker.ts

```diff
@@ -9,83 +9,322 @@
+  }
+
+  public delete(): void {
+    this._pvDownsamplerDelete(this._objectAddress);
+  }
+
+  public getrequiredNumInputSamples(numSample: number): number {
+    return this._pvDownsamplerConvertNumSamplesToInputSampleRate(
+      this._objectAddress,
+      numSample,
+    );
+  }
+
+  private static arrayBufferToStringAtIndex(
+    arrayBuffer: Uint8Array,
+    index: number,
+  ): string {
+    let stringBuffer = '';
+    let indexBuffer = index;
+    while (true) {
+      if (arrayBuffer[indexBuffer] === 0) {
+        break;
+      }
+      stringBuffer += String.fromCharCode(arrayBuffer[indexBuffer++]);
+    }
+    return stringBuffer;
+  }
+
+  private static base64ToUint8Array(base64String: string): Uint8Array {
+    const base64StringDecoded = atob(base64String);
+    const binaryArray = new Uint8Array(base64StringDecoded.length);
+    for (let i = 0; i < base64StringDecoded.length; i++) {
+      binaryArray[i] = base64StringDecoded.charCodeAt(i);
+    }
+    return binaryArray;
+  }
+}
 
 const PV_FRAME_LENGTH = 512;
 const PV_SAMPLE_RATE = 16000;
-const S_INT16_MAX = 32767;
+const PV_FILTER_ORDER = 50;
 
-let _inputSampleRate: number;
-let _outputSampleRate: number;
-let _frameLength: number;
-let _inputBuffer: Array<number> = [];
+let _downsampler: Downsampler;
+let _outputframeLength: number;
+let _inputBuffer: Int16Array;
+let _outputBuffer: Int16Array;
 
 let _audioDumpActive: boolean;
 let _audioDumpBuffer: Int16Array;
 let _audioDumpBufferIndex: number;
-let _audioDumpNumFrames: number;
+let _audioDumpNumSamples: number;
 
-function init(
+async function init(
   inputSampleRate: number,
   outputSampleRate: number = PV_SAMPLE_RATE,
   frameLength: number = PV_FRAME_LENGTH,
-): void {
-  _inputSampleRate = inputSampleRate;
-  _outputSampleRate = outputSampleRate;
-  _frameLength = frameLength;
+): Promise<void> {
+  console.assert(Number.isInteger(inputSampleRate));
+  console.assert(Number.isInteger(outputSampleRate));
+  console.assert(Number.isInteger(frameLength));
+
+  _outputframeLength = frameLength;
 
-  console.assert(Number.isInteger(_inputSampleRate));
-  console.assert(Number.isInteger(_outputSampleRate));
-  console.assert(Number.isInteger(_frameLength));
+  _downsampler = await Downsampler.create(
+    inputSampleRate,
+    outputSampleRate,
+    PV_FILTER_ORDER,
+    _outputframeLength,
+  );
 
-  _inputBuffer = [];
+  postMessage(
+    {
+      command: 'ds-ready',
+    },
+    undefined as any,
+  );
 }
 
 function startAudioDump(durationMs: number = 3000): void {
-  _audioDumpNumFrames = durationMs * (PV_FRAME_LENGTH / PV_SAMPLE_RATE);
+  _audioDumpNumSamples = (durationMs * PV_SAMPLE_RATE) / 1000;
   _audioDumpActive = true;
   _audioDumpBufferIndex = 0;
-  _audioDumpBuffer = new Int16Array(_audioDumpNumFrames * _frameLength);
+  _audioDumpBuffer = new Int16Array(_audioDumpNumSamples);
 }
 
 function processAudio(inputFrame: Float32Array): void {
+  console.assert(inputFrame.constructor === Float32Array);
```

Comment:
good runtime type check here. go ahead and throw an Error if the condition is not met.